### PR TITLE
refactor: make modules_mapping a regular rule

### PR DESCRIPTION
**Pull request by f0rmiga**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Does not include precompiled binaries, eg. `.par` files. See [CONTRIBUTING.md](https://github.com/bazelbuild/rules_python/blob/master/CONTRIBUTING.md) for info
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What has changes

The current `modules_mapping` is a repository rule. This changeset moves it to be a regular rule, consuming the wheels from pip_install instead of re-downloading and building (when necessary) them twice.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

Since the Gazelle plugin is not contained in any release, I don't consider it to be a breaking change.

### CODE REVIEW

File: gazelle/README.md [hunk 1]
```diff
@@ -61,7 +57,7 @@ gazelle_python_manifest(
     name = "gazelle_python_manifest",
     # The @modules_map refers to the name we gave in the modules_mapping
     # rule in the WORKSPACE
```
Comment (kormide) [inline gazelle/README.md:59]: Reference to the old repository name.

**Comment by kormide** [comment]
Was there a reason why `modules_mapping` was originally a repository rule rather than a _rule_ rule?

**Comment by f0rmiga** [comment]
@kormide In the early days, the modules mapping was considerably faster compared to depending on all the wheels from `pip_install`. It was important because we were generating it on the fly (not checked-in as part of the gazelle manifest like we do now). This is not true anymore.

File: gazelle/README.md [hunk 2]
```diff
@@ -23,22 +20,12 @@ Add this to your `WORKSPACE`:
...
-)
 ```
 
-Next, we'll make a pair of targets for consuming that `modules_mapping` we
-fetched, and writing it as a manifest file for Gazelle to read.
+Next, we'll fetch metadata about your Python dependencies, so that gazelle can
+determine which package a given import statement comes from. This is provided
+by the `modules_mapping` rule. We'll make a target for consuming this
+`modules_mapping`, and writing it as a manifest file for Gazelle to read.
 This is checked into the repo for speed, as it takes some time to calculate
```
Comment (alexeagle) [inline gazelle/README.md:29]: Is there still a file checked into the repo for speed?

**Review by alexeagle — APPROVED** [review]

File: gazelle/README.md (hunk 2 above)
Comment (f0rmiga) [reply gazelle/README.md:29]: Yes

File: gazelle/README.md (hunk 1 above)
Comment (f0rmiga) [reply gazelle/README.md]: Fixed in https://github.com/bazelbuild/rules_python/pull/578/commits/fa6a7cf2f3b11eed1a0e382a0c1823f2d7637cfe.

**Comment by corypaik** [comment]
I was working on this a bit earlier since the previous version didn't work with `--find-links` and thought it could make sense to generate a `modules_mapping.json` for each wheel individually. I've refactored what I had to work with this rule and drafted a change that runs the generator once per wheel then merges the result. It seems to address the issue of having to recompute the module mapping for all dependencies each time one is updated. 

Not really sure of a good way to benchmark it beyond rough timings and checking for cached actions, but it seems like it should be more efficient. Could also combine the json merger with the yaml generator if having them separate adds too much overhead, but I wasn't sure if there was another use for the merged json file beyond making the yaml one. I pushed the changes [here](https://github.com/corypaik/rules_python/tree/simplify-modules-mapping).

**Comment by f0rmiga** [comment]
@corypaik - interesting approach. We have been using it on a large project and haven't felt the need to do this performance improvement since the testing to check whether the checked-in manifest is up to date or not doesn't require recomputing the whole file.

The most expensive part of updating the modules mapping is downloading and, in some cases, compiling native modules. The tree traversal to generate the mappings is usually super fast and I don't think the performance improvement would be noticeable?

**Comment by aignas** [comment]
I think making this a simple rule also fixed #574 because the same fetch packages can be now used here. So thanks for this change!

**Pull request merged**